Source/WTF/wtf/text/CStringView.h

```diff
@@ -0,0 +1,177 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include <span>
+#include <string>
+#include <type_traits>
+#include <wtf/Compiler.h>
+#include <wtf/ForbidHeapAllocation.h>
+#include <wtf/Forward.h>
+#include <wtf/HashFunctions.h>
+#include <wtf/StdLibExtras.h>
+#include <wtf/text/ASCIILiteral.h>
+#include <wtf/text/SuperFastHash.h>
+
+namespace WTF {
+
+class PrintStream;
+class String;
+
+class CStringView final {
+    WTF_FORBID_HEAP_ALLOCATION;
+public:
+    static CStringView fromLatin1(const char* string LIFETIME_BOUND)
+    {
+        if (!string)
+            return CStringView();
+        return CStringView(unsafeMakeSpan(static_cast<const LChar*>(static_cast<const void*>(string)), std::char_traits<char>::length(string) + 1));
+    }
+
+    static CStringView fromNullTerminatedLatin1(std::span<const char> spanWithNullTerminator LIFETIME_BOUND)
+    {
+        RELEASE_ASSERT(spanWithNullTerminator.size() && spanWithNullTerminator[spanWithNullTerminator.size() - 1] == '\0');
+        return CStringView(byteCast<LChar>(spanWithNullTerminator));
+    }
+
+    static CStringView fromNullTerminatedLatin1(std::span<const LChar> spanWithNullTerminator LIFETIME_BOUND)
+    {
+        RELEASE_ASSERT(spanWithNullTerminator.size() && spanWithNullTerminator[spanWithNullTerminator.size() - 1] == '\0');
+        return CStringView(spanWithNullTerminator);
+    }
+
+    static CStringView fromNullTerminatedUChar(std::span<const UChar> spanWithNullTerminator LIFETIME_BOUND)
+    {
+        RELEASE_ASSERT(spanWithNullTerminator.size() && spanWithNullTerminator[spanWithNullTerminator.size() - 1] == '\0');
+        return CStringView(spanWithNullTerminator);
+    }
+
+    WTF_EXPORT_PRIVATE void dump(PrintStream& out) const;
+
+    CStringView() = default;
+    constexpr CStringView(std::nullptr_t)
+        : CStringView()
+    { }
+    CStringView(ASCIILiteral literal LIFETIME_BOUND)
+    {
+        if (!literal.length())
+            return;
+        m_charactersWithNullTerminator = literal.spanIncludingNullTerminator().data();
+        m_lengthWithNullTerminator = literal.spanIncludingNullTerminator().size();
+    }
+
+    unsigned hash() const;
+    bool isNull() const { return !m_lengthWithNullTerminator; }
+
+    const char* characters() const LIFETIME_BOUND { return static_cast<const char*>(m_charactersWithNullTerminator); }
```

Comment:
I don't think we want a bounds-unsafe and encoding-forgetting accessor if we can avoid it. Is this accessor absolutely necessary?